flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java

```diff
@@ -17,156 +17,489 @@
+		this.streams = checkNotNull(streams);
 		this.configProps = checkNotNull(configProps);
-		this.assignedShardsWithStartingSequenceNum = new HashMap<>();
-		for (KinesisStreamShard shard : assignedShards) {
-			assignedShardsWithStartingSequenceNum.put(shard, SentinelSequenceNumber.SENTINEL_SEQUENCE_NUMBER_NOT_SET.toString());
+		this.sourceContext = checkNotNull(sourceContext);
+		this.checkpointLock = checkNotNull(checkpointLock);
+		this.runtimeContext = checkNotNull(runtimeContext);
+		this.deserializationSchema = checkNotNull(deserializationSchema);
+
+		this.error = error;
+		this.pendingShards = pendingShardsQueue;
+		this.subscribedShardsState = subscribedShardsState;
+
+		this.shardDiscovererAndSubscriberExecutor =
+			createShardDiscovererAndSubscriberThreadPool(runtimeContext.getTaskNameWithSubtasks());
+		this.shardConsumersExecutor =
+			createShardConsumersThreadPool(runtimeContext.getTaskNameWithSubtasks());
+
+		this.isRestoredFromCheckpoint = (restoredState != null && restoredState.entrySet().size() != 0);
+
+		// if there is state to restore from last checkpoint, we seed them as initially discovered shards
+		if (isRestoredFromCheckpoint) {
+			seedPendingShardsWithRestoredState(restoredState, this.pendingShards);
 		}
-		this.taskName = taskName;
-		this.error = new AtomicReference<>();
 	}
 
 	/**
-	 * Advance a shard's starting sequence number to a specified value
+	 * Starts the fetcher. After starting the fetcher, it can only
+	 * be stopped by calling {@link KinesisDataFetcher#shutdownFetcher()}.
 	 *
-	 * @param streamShard the shard to perform the advance on
-	 * @param sequenceNum the sequence number to advance to
+	 * @throws Exception the first error or exception thrown by the fetcher or any of the threads created by the fetcher.
 	 */
-	public void advanceSequenceNumberTo(KinesisStreamShard streamShard, String sequenceNum) {
-		if (!assignedShardsWithStartingSequenceNum.containsKey(streamShard)) {
-			throw new IllegalArgumentException("Can't advance sequence number on a shard we are not going to read.");
+	public void runFetcher() throws Exception {
+
+		// atomic operation to startup the fetcher and create the shard discoverer and subscriber.
+		synchronized (fetcherShutdownLock) {
+
+			// this flag will be false if the shutdown procedure was
+			// executed first; if so, we return without doing anything.
+			if (!running) {
+				return;
+			}
+
+			if (LOG.isInfoEnabled()) {
+				LOG.info("Subtask {} is starting the shard discoverer ...", runtimeContext.getIndexOfThisSubtask());
+			}
+			shardDiscovererAndSubscriberExecutor.submit(new ShardDiscoverer<>(this));
+
+			if (LOG.isInfoEnabled()) {
+				LOG.info("Subtask {} is starting the shard subscriber ...", runtimeContext.getIndexOfThisSubtask());
+			}
+			shardDiscovererAndSubscriberExecutor.submit(new ShardSubscriber<>(this));
+
+			// after startup, we wait util either shutdownFetcher() or stopWithError() is called
+			fetcherShutdownLock.wait();
```

Comment:
I think the `shutdownFetcher()` can never call notifyAll because this method is never releasing the lock on `fetcherShutdownLock`.

What happens instead is that Flink is interrupting the thread holding the lock. This interrupt will cause an InterruptedException, which is not catched at all.

I don't think this waiting mechanism is necessary here at all. Maybe we can use the `awaitTermination()` for that (but maybe there is an even better solution available ... I'll continue with the review)